### PR TITLE
MINOR: Update unit/integration tests to work with the IBM Semeru JDK

### PR DESCRIPTION
**Pull request by jonathan-albrecht-ibm**
The IBM Semeru JDK use the OpenJDK security providers instead of the IBM
security providers so test for the OpenJDK classes first where possible
and test for Semeru in the java.runtime.name system property otherwise.

Use a real but empty KafkaThread object in KafkaProducerTest instead of
a mock because the Semeru Thread class implementation has some extra
checks that are called and fail when the mock is used.

Tested with IBM and Temurin JDKs. Added a unit test for the new Java.isIbmSemeru() method.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Review by Kvicii — APPROVED** [review]

**Comment by jonathan-albrecht-ibm** [comment]
Thanks for the review @Kvicii. The CI seems to have failed due to a server error and looks stuck. Could the build be restarted?

**Comment by jonathan-albrecht-ibm** [comment]
Hi @Kvicii, I have rebased this PR to trigger the tests. It looks like everything that ran passed except for 2 instances of
```
Build / JDK 8 and Scala 2.12 / testFencedLeaderRecovery – org.apache.kafka.connect.integration.ExactlyOnceSourceIntegrationTest
```
From a quick look at other PR tests, I think these are just flakey eg. https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-12396/3/pipeline/14

Would it be possible to merge this PR? Let me know if there is anything else I need to do.

**Comment by jonathan-albrecht-ibm** [comment]
I have rebased this PR to latest trunk. I don't think the tests that failed are related to these changes. Would it be possible for someone to review?

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java
Comment (cadonna) [inline clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: Is it possible to set up the mock in such a way, so that it can be used? 
Are there native methods involved? That woulod be a reason to not use a mock since Mockito cannot mock native methods.

File: core/src/test/scala/kafka/security/minikdc/MiniKdc.scala
Comment (cadonna) [inline core/src/test/scala/kafka/security/minikdc/MiniKdc.scala]: Why do you not use a method `Java.isIbmJdkSemeru` here similar to what you did for the clients? I think that would improve readability and avoid the comment.

File: core/src/test/scala/unit/kafka/utils/JaasTestUtils.scala
Comment (cadonna) [inline core/src/test/scala/unit/kafka/utils/JaasTestUtils.scala]: See my comment above.

**Review by cadonna — COMMENTED** [review]
@jonathan-albrecht-ibm Thank you for the PR!

Here my feedback!

Since I am not familiar with this code, would be good if @ijuma and @hachikuji could have a look at this PR.

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java
Comment (jonathan-albrecht-ibm) [reply clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: I have run the test again without this change and it looks like it is no longer needed. I see that mockito has been upgraded to newer releases a few times since this PR was created so I'm guessing that has fixed this problem.

I'll push an amended commit to remove this change but let me know if the kafka developers prefer adding a new commit to remove it instead.

File: core/src/test/scala/kafka/security/minikdc/MiniKdc.scala
Comment (jonathan-albrecht-ibm) [reply core/src/test/scala/kafka/security/minikdc/MiniKdc.scala]: The idea was to just check for the classes directly but you're definitely right that `Java.isIbmJdkSemeru` is more readable. I have a new commit to make this change and the other one below, but I'll hold off on pushing it to see if any more feedback from @ijuma or @hachikuji.

File: core/src/test/scala/unit/kafka/utils/JaasTestUtils.scala
Comment (jonathan-albrecht-ibm) [reply core/src/test/scala/unit/kafka/utils/JaasTestUtils.scala]: See reply above.

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java
Comment (jonathan-albrecht-ibm) [reply clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: Done. Amended original commit.

File: core/src/test/scala/kafka/security/minikdc/MiniKdc.scala
Comment (jonathan-albrecht-ibm) [reply core/src/test/scala/kafka/security/minikdc/MiniKdc.scala]: Done

File: core/src/test/scala/unit/kafka/utils/JaasTestUtils.scala
Comment (jonathan-albrecht-ibm) [reply core/src/test/scala/unit/kafka/utils/JaasTestUtils.scala]: Done

**Comment by jonathan-albrecht-ibm** [comment]
TFTR @cadonna, I've updated the changes according to your feedback.

**Review by mimaison — APPROVED** [review]
LGTM, thanks for the PR!

**Comment by jonathan-albrecht-ibm** [comment]
Thanks @mimaison!

File: clients/src/test/java/org/apache/kafka/common/utils/JavaTest.java
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/common/utils/JavaTest.java:61]: Looks like we don't clear the set property, won't that pollute the rest of the test suite that runs in that jvm?

File: clients/src/main/java/org/apache/kafka/common/utils/Java.java
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/utils/Java.java:47]: Instead of having this with the logic in various places, can we not treat IBM Semeru as plain openjdk? If so, we could change the check for IBM JDK to exclude IBM Semeru.

File: clients/src/test/java/org/apache/kafka/common/utils/JavaTest.java
Comment (jonathan-albrecht-ibm) [reply clients/src/test/java/org/apache/kafka/common/utils/JavaTest.java:61]: The original property values are saved and restored in the `before()` and `after()` methods (annotated with `@BeforeEach` and `@AfterEach`) so I think each test method will start with the JVM's property values. Let me know if I'm missing something.

Comment (ijuma) [reply clients/src/test/java/org/apache/kafka/common/utils/JavaTest.java:61]: Thanks for the quick response. Yes, you're right, I had missed that.

File: clients/src/main/java/org/apache/kafka/common/utils/Java.java
Comment (jonathan-albrecht-ibm) [reply clients/src/main/java/org/apache/kafka/common/utils/Java.java:47]: I don't know if we can treat the IBM Semeru JDK as openjdk in all places where `isIbmJdk()` is being used. The changes here fix the places where the IBM security classes were being incorrectly specified for Semeru. I could look at the other places but since there aren't tests that are failing or obvious regression, I would need to investigate more.

**Pull request merged**
